Pymod msvc

klayoutmatthias opened the pull request:
Hi Thomas,

the main changes of interest are those in setup.py. Windows has some different concepts, i.e.
 * import libraries: those are produced along with a DLL, but placed in a different folder. Instead of the DLL, these files have to be linked against
 * explicit symbol exports (requires to use "Library" instead of "Extension" because for extensions, setuptools tries to export the PyInit symbol - which fails)
 * separate paths for Debug and Release builds

The setup.py script changed significantly, but I hope it won't break MacOS builds.

Thanks

Matthias

klayoutmatthias committed:
Activated stream unit tests.

klayoutmatthias committed:
Fixed #185 for pymod branch

klayoutmatthias committed:
Updated Changelog (merge from master)

klayoutmatthias committed:
Fixed build on Windows (tl::InputPipe::wait was missing)

klayoutmatthias committed:
setup.py changes for MSVC builds

- Using "Library" rather than "Extension" for building shared libs
  (does not try to export extension-specific symbols)
- Linking against import libs
- Using os.path.join consistently for backslash/slash difference

klayoutmatthias committed:
Updated version to check CD

thomaslima commented:
This Pull request breaks the build under mac and linux (at least) because distutils picks different naming for .so objects between libraries vs extensions. e.g. `lib_gsi.cpython-37m-darwin.so` instead of `_gsi.cpython-37m-darwin.so`. I spent a lot of time looking into their undocumented source code but this is hardcoded in distutils and there is no way we can simply extract the proper naming from the setup.py script. We have to indeed manually add that "lib" prefix to the generated filename. See the errors below.

https://travis-ci.org/lightwave-lab/klayout/jobs/443417666

I can't push to your branch because I don't have dev access to your repo. I'd like to have a clean merge if possible.

klayoutmatthias commented:
Hi Thomas,

thanks for looking into this. And you're perfectly right. I myself found distutils to have too many such hardcoded rules for my taste. Still I think going to "Library" instead of "Extension" is the right way because that's what the _tl ... libs are.

I have added you as collaborator my be repo, so you should be able to push the changes. 

Thanks and best regards,

Matthias

klayoutmatthias committed:
Fixed setup.py for Linux (and maybe for MacOS too)

- Added "lib" prefix for libraries
- Forced setuptools to produce shared objects for the
  libraries - otherwise it will produce static libs

thomaslima committed:
fixing dbpi library names

klayoutmatthias commented:
Hi Thomas,

sorry for the conflict, if there was one ...

I tried to fix the MacOS build too, but my patch was essentially using "dylib" libtype for Darwin:

```
-setuptools.command.build_ext.libtype = "shared"
+if platform.system() == "Darwin":
+    setuptools.command.build_ext.libtype = "dylib"
+else:
+    setuptools.command.build_ext.libtype = "shared"
```

I don't know whether that has any side effects so maybe yours is better.

Let's see what Travis says :-)

Matthias

thomaslima commented:
I see. My patch works for python3 apparently. The annoying thing is that distutils for python2 and python3 have different behaviors. Somehow it's using dylib as extension instead of .so. I'm debugging.

thomaslima committed:
attempt to solve the libname_of computation for all platforms

thomaslima commented:
My solution attempts to actually using distutils to calculate the right filename, so we're future-proof. Your solution does have the merit of being short though...

Travis is having a few internal issues on the MacOS builds but I think this branch is good to be merged. Thanks for your help!

My next PR will be on building wheels for the manylinux docker image. Stay tuned.

thomaslima commented:
Matthias, please merge this on your repo. I am mobile now and can’t submit pull requests. This PR will close automatically once it detects a merge.

thomaslima commented:
Nevermind I merged.

klayoutmatthias commented:
Thanks for your efforts ... I merged the whole thing into pymod now.

Best regards,

Matthias
